File: packages/tools/src/redpilot_tools/utils/path_safety.py
"""Path safety utilities — containment checks for filesystem tool operations.

Every filesystem tool (list_directory, read_file, write_file, edit_file) must
validate that the requested path is contained within the engagement's
scratch/evidence root directory. This module provides the canonical
containment check that rejects:

- Path traversal (``../`` components escaping the root)
- Absolute paths outside the root
- Symlink escapes (resolves all symlinks before checking)
- Null bytes and other suspicious patterns
"""

from __future__ import annotations

import os
from pathlib import Path


class PathEscapeError(ValueError):
    """Raised when a path attempts to escape the allowed root directory."""


_SUSPICIOUS_PATTERNS = [
    b"\x00",  # Null bytes
]


def resolve_safe_path(
    requested_path: str,
    root_dir: str,
    *,
    must_exist: bool = False,
) -> Path:
    """Resolve *requested_path* relative to *root_dir* and validate containment.

    Steps:
    1. Reject paths with suspicious content (null bytes, etc.)
    2. Resolve the requested path relative to root_dir
    3. Resolve symlinks (``Path.resolve()``)
    4. Verify the resolved path starts with the resolved root_dir
    5. If *must_exist*, verify the path actually exists on disk

    Args:
        requested_path: The user/LLM-supplied path string. May be relative
            or absolute.
        root_dir: The allowed root directory. All operations must stay
            within this directory tree.
        must_exist: If True, raises ``FileNotFoundError`` if the resolved
            path does not exist on disk.

    Returns:
        The resolved, containment-verified ``Path`` object.

    Raises:
        PathEscapeError: If the path attempts to escape the root directory
            or contains suspicious patterns.
        FileNotFoundError: If *must_exist* is True and the path doesn't exist.
    """
    # 1. Check for suspicious binary content
    _check_suspicious_patterns(requested_path)

    # 2. Resolve both root and requested paths
    root = Path(root_dir).resolve()
    resolved = (root / requested_path).resolve()

    # 3. Verify containment
    try:
        resolved.relative_to(root)
    except ValueError:
        msg = (
            f"Path '{requested_path}' resolves to '{resolved}' "
            f"which is outside the allowed root directory '{root}'"
        )
        raise PathEscapeError(msg) from None

    # 4. Check must_exist
    if must_exist and not resolved.exists():
        msg = f"Path '{requested_path}' does not exist (resolved to '{resolved}')"
        raise FileNotFoundError(msg)

    return resolved


def resolve_safe_path_allow_missing(
    requested_path: str,
    root_dir: str,
) -> Path:
    """Like ``resolve_safe_path`` but does not require the path to exist.

    Useful for write operations where the file is being created for the
    first time, or for edit operations where we check existence separately
    with a clearer error message.

    Args:
        requested_path: The user/LLM-supplied path string.
        root_dir: The allowed root directory.

    Returns:
        The resolved, containment-verified ``Path`` object.

    Raises:
        PathEscapeError: If the path attempts to escape the root.
    """
    return resolve_safe_path(requested_path, root_dir, must_exist=False)


def resolve_safe_path_must_exist(
    requested_path: str,
    root_dir: str,
) -> Path:
    """Like ``resolve_safe_path`` but requires the path to already exist.

    Useful for read and edit operations where the file must already
    exist on disk.

    Args:
        requested_path: The user/LLM-supplied path string.
        root_dir: The allowed root directory.

    Returns:
        The resolved, containment-verified ``Path`` object.

    Raises:
        PathEscapeError: If the path attempts to escape the root.
        FileNotFoundError: If the path does not exist.
    """
    return resolve_safe_path(requested_path, root_dir, must_exist=True)


def _check_suspicious_patterns(path_str: str) -> None:
    """Check for suspicious patterns in the path string.

    Raises:
        PathEscapeError: If a suspicious pattern is found.
    """
    encoded = path_str.encode("utf-8", errors="replace")
    for pattern in _SUSPICIOUS_PATTERNS:
        if pattern in encoded:
            msg = f"Path contains suspicious pattern (null byte): '{path_str}'"
            raise PathEscapeError(msg)

File: apps/tui/src/utils/path-safety.ts
/** Path safety — containment checks for local filesystem operations.
 *
 * Every filesystem tool (list_directory, read_file, write_file, edit_file)
 * must validate that the requested path stays within the sandbox root.
 *
 * The sandbox root defaults to ~/.redpilot/scratch/ and can be overridden
 * via the REDPILOT_SCRATCH_DIR env var.
 *
 * Uses Path.resolve() semantics: resolving a relative path joins it to the
 * root, while resolving an absolute path replaces the root entirely.
 * The startsWith(root) check catches all escape attempts including:
 *   - ../ traversal resolving above the root
 *   - Absolute paths like /etc/passwd that replace the root
 *   - Symlink escapes (handled at the OS level by resolve())
 */

import { existsSync, mkdirSync } from "node:fs";
import { homedir } from "node:os";
import { resolve } from "node:path";

export class PathEscapeError extends Error {
  constructor(path: string, root: string) {
    super(
      `Path '${path}' escapes the sandbox root '${root}'. ` +
      "Path traversal and absolute paths outside the sandbox are not allowed.",
    );
    this.name = "PathEscapeError";
  }
}

/** Get the sandbox root directory. */
export function getSandboxRoot(): string {
  const envDir = typeof process !== "undefined"
    ? process.env.REDPILOT_SCRATCH_DIR
    : undefined;
  return envDir ?? resolve(homedir(), ".redpilot", "scratch");
}

/** Ensure the sandbox root directory exists. */
export function ensureSandboxRoot(): string {
  const root = getSandboxRoot();
  mkdirSync(root, { recursive: true });
  return root;
}

/**
 * Resolve a path relative to the sandbox root and validate containment.
 *
 * Uses a single guard: resolve the path, then verify the resolved path
 * starts with the sandbox root. This catches:
 *   - ``../etc/passwd`` → resolves above root → rejected by startsWith
 *   - ``/etc/passwd`` → absolute replaces root → rejected by startsWith
 *   - ``subdir/../../../etc`` → normalizes above root → rejected by startsWith
 *
 * @param requestedPath - The user/LLM-supplied path (relative or absolute)
 * @param sandboxRoot - The allowed root directory (defaults to getSandboxRoot())
 * @returns The resolved, contained absolute path
 * @throws PathEscapeError if the path escapes the sandbox
 */
export function resolveSafePath(
  requestedPath: string,
  sandboxRoot?: string,
): string {
  const root = resolve(sandboxRoot ?? getSandboxRoot());

  // Reject null bytes
  if (requestedPath.includes("\0")) {
    throw new PathEscapeError(requestedPath, root);
  }

  // Resolve: Path.resolve() normalizes .. and treats absolute paths as-is
  const resolved = resolve(root, requestedPath);

  // Single guard: resolved path must start with the root.
  // Use root + "/" separator to prevent bypass: e.g. "/dir/scratch-other/file"
  // starts with "/dir/scratch" but is outside the sandbox.
  const rootGuard = root.endsWith("/") ? root : root + "/";
  if (resolved !== root && !resolved.startsWith(rootGuard)) {
    throw new PathEscapeError(requestedPath, root);
  }

  return resolved;
}

/**
 * Like resolveSafePath, but also verifies the path exists on disk.
 * @throws PathEscapeError if path escapes
 * @throws Error if path doesn't exist
 */
export function resolveSafePathMustExist(
  requestedPath: string,
  sandboxRoot?: string,
): string {
  const resolved = resolveSafePath(requestedPath, sandboxRoot);
  if (!existsSync(resolved)) {
    throw new Error(
      `Path '${requestedPath}' does not exist (resolved to '${resolved}')`,
    );
  }
  return resolved;
}
